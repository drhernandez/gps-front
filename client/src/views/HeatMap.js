// import PropTypes from "prop-types";
import React from "react";
import {
  Container,
  Row,
  Col,
  Card,
  CardHeader,
  CardBody,
  FormSelect,
  Form,
  Alert
} from "shards-react";
import constants from "../utils/Constants";
import PageTitle from "./../components/common/PageTitle";
import Gmap from "./../components/common/maps/Gmap";
import to from "await-to-js";
import { VehiclesService } from "../api/services"
import store from "../redux/store";

const alerts = {
  generic: {
    type: constants.Themes.ERROR,
    message: "Hubo un error al intentar cargar los datos. Inténtelo de nuevo"
  },
  trackingsNotFound: {
    type: constants.Themes.WARNING,
    message: "No se han registrado datos para este vehículo"
  }
}

export default class HeatMap extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      vehicles: [],
      trackings: [],
      markers: [],
      center: null,
      zoom: null,
      alert: {
        visible: false,
        body: {}
      }
    }
  };

  componentDidMount() {
    this.loadVehicles();
  }

  async loadVehicles() {
    const userId = store.getState().userInfo.id;
    const [err, response] = await to(VehiclesService.searchVehicles(userId));
    if (!err && response.paging.total) {
      this.setState({
        vehicles: response.data
      });
    }
  }
  
  async getTrackings(vehicleID) {
    const [err, trackings] = await to(VehiclesService.getTrackings(vehicleID));
    if (!err && trackings.length) {
      this.setState({
        trackings: trackings,
        alert: {
          visible: false,
          body: {}
        }
      });
    }
    else {
      this.setState({
        trackings: [],
        markers: [],
        center: null,
        zoom: null,
        alert: {
          visible: true,
          body: !err && !trackings.length ? alerts.trackingsNotFound : alerts.generic
        }
      })
    }
  }

  handleDeviceOnChange(event) {
    event.persist();
    this.getTrackings(event.target.value);
  }

  render() {
    return (
      <Container fluid className="main-content-container px-4">
        {/* Page Header */}
        <Row noGutters className="page-header py-4">
          <Col>
            <PageTitle title="Ubicación actual" subtitle="Mapa" className="text-sm-left mb-3 col-sm-12" />
          </Col>
          <Col>
            <Alert open={this.state.alert.visible} className="m-0" theme={this.state.alert.body.type}>
              {this.state.alert.body.message}
            </Alert>
          </Col>
        </Row>

        <Row>
          <Col>
            <Card small className="mb-4">
              <CardHeader>
                <Form className="add-new-post">
                  <FormSelect id="feInputState" defaultValue="default" onChange={(event) => this.handleDeviceOnChange(event)}>
                    <option value="default" disabled>Elija un vehículo...</option>
                    {this.state.vehicles.map((vehicle, idx) => (
                      <option key={vehicle.id} value={vehicle.id}>{`${vehicle.brand} ${vehicle.brand_line} - ${vehicle.plate}`}</option>
                    ))}
                  </FormSelect>
                </Form>
              </CardHeader>
              <CardBody>
                <Gmap
                  isHeatMapLayerShown
                  containerElement={<div style={{ height: '450px' }} />}
                  mapElement={<div style={{ height: `100%` }} />}
                  // zoom={this.state.zoom}
                  // markers={this.state.markers}
                  center={this.state.center}
                  trackings={this.state.trackings}
                />
              </CardBody>
            </Card>
          </Col>
        </Row>
      </Container>
    );
  }
}

// function getDefaultIcon() {
//   const svg = `<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M12 2C8.13 2 5 5.13 5 9c0 5.25 7 13 7 13s7-7.75 7-13c0-3.87-3.13-7-7-7zm0 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z" fill="#ea4234"/><path d="M0 0h24v24H0z" fill="none"/></svg>`
//   return { url: 'data:image/svg+xml;charset=UTF-8,' + encodeURIComponent(svg), scaledSize: { width: 40, height: 40 } }
// }