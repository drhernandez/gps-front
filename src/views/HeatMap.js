/*global google*/
import React from "react";
// import PropTypes from "prop-types";
import {
  Container,
  Row,
  Col,
  Card,
  CardHeader,
  CardBody,
  FormSelect,
  Form,
} from "shards-react";
import PageTitle from "./../components/common/PageTitle";
import Gmap from "./../components/maps/Gmap";
import { Marker } from "./../components/maps/Marker";
//services
import UsersService from "./../api/services/usersService";
import VehiclesService from "./../api/services/vehiclesService";


export default class HeatMap extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      vehicles: [],
      trackings: [],
      markers: [],
      center: null,
      zoom: null
    }
  };

  async loadDevices() {
    const usersService = new UsersService();
    const vehicles = await usersService.getVehiclesByUserID(10);
    this.setState({
      vehicles: vehicles
    });
  }

  async getCurrentLocation(vehicleID) {
    const vehiclesServices = new VehiclesService();
    const location = await vehiclesServices.getCurrentLocation(vehicleID);
    if (location !== null || location !== undefined) {
      const marker = new Marker(location.id, location.label, undefined, google.maps.Animation.DROP, location.lat, location.lng);
      this.setState({
        trackings: [location],
        markers: [marker],
        center: marker.position,
        zoom: 15
      })
    }
  }

  handleDeviceOnChange(event) {
    event.persist();
    this.getCurrentLocation(event.target.value);
  }

  componentWillMount() {
    this.loadDevices();
  }

  render() {
    return (
      <Container fluid className="main-content-container px-4">
        {/* Page Header */}
        <Row noGutters className="page-header py-4">
          <PageTitle title="Where is my car?!" subtitle="Dashboard" className="text-sm-left mb-3 col-sm-12" />
        </Row>

        <Row>
          <Col>
            <Card small className="mb-4">
              <CardHeader>
                <Form className="add-new-post">
                  <FormSelect id="feInputState" defaultValue="default" onChange={(event) => this.handleDeviceOnChange(event)}>
                    <option value="default" disabled>Elija un veh√≠culo...</option>
                    {this.state.vehicles.map((vehicle, idx) => (
                      <option key={vehicle.device_id} value={vehicle.id}>{`${vehicle.type} - ${vehicle.plate}`}</option>
                    ))}
                  </FormSelect>
                </Form>
              </CardHeader>
              <CardBody>
                <Gmap
                  googleMapURL="http://maps.google.com/maps/api/js?key=AIzaSyAAx3DLDKUqNQOEp8zi3G2xEuAYfgZoPTo"
                  loadingElement={<div style={{ height: `100%` }} />}
                  containerElement={<div style={{ height: '450px' }} />}
                  mapElement={<div style={{ height: `100%` }} />}
                  zoom={this.state.zoom}
                  markers={this.state.markers}
                  center={this.state.center}
                />
              </CardBody>
            </Card>
          </Col>
        </Row>
      </Container>
    );
  }
}

// function getDefaultIcon() {
//   const svg = `<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M12 2C8.13 2 5 5.13 5 9c0 5.25 7 13 7 13s7-7.75 7-13c0-3.87-3.13-7-7-7zm0 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z" fill="#ea4234"/><path d="M0 0h24v24H0z" fill="none"/></svg>`
//   return { url: 'data:image/svg+xml;charset=UTF-8,' + encodeURIComponent(svg), scaledSize: { width: 40, height: 40 } }
// }