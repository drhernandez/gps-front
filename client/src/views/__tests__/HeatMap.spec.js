import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
const mockStore = configureStore([]);

jest.mock("../../api/services/vehiclesService", () => ({ searchVehicles: jest.fn(), searchTrackings: jest.fn() }));
import VehiclesService from '../../api/services/vehiclesService';

import HeatMap from '../HeatMap';

function flushPromises() {
  return new Promise(resolve => setImmediate(resolve));
}

let store;
let searchMock;
let trackingsMock;

describe('HeatMap component', () => {
  beforeEach(async () => {
    jest.resetModules();
    jest.useFakeTimers()

    store = mockStore({
      userInfo: {
        id: 1
      },
    });

    searchMock = {
      data: [
        {
          id: 1,
          brand: "Frod",
          brand_line: "Fiesta",
          plate: "AA111AA",
          device: {
            id: 3
          }
        }
      ],
      paging: {
        total: 1
      }
    }

    trackingsMock = {
      data: [
        {
          "id": 1,
          "device_id": 1,
          "lat": -31.422108,
          "lng": -64.18643
        },
        {
          "id": 2,
          "device_id": 1,
          "lat": -31.422218,
          "lng": -64.18578
        },
        {
          "id": 3,
          "device_id": 1,
          "lat": -31.422731,
          "lng": -64.184845
        },
        {
          "id": 4,
          "device_id": 1,
          "lat": -31.42284,
          "lng": -64.183945
        },
        {
          "id": 5,
          "device_id": 1,
          "lat": -31.423098,
          "lng": -64.18351
        },
        {
          "id": 6,
          "device_id": 1,
          "lat": -31.423016,
          "lng": -64.183205
        }
      ],
      paging: {
        total: 6
      }
    }

    await VehiclesService.searchVehicles.mockImplementation(() => {
      return Promise.resolve(searchMock);
    });
    await VehiclesService.searchTrackings.mockImplementation(() => {
      return Promise.resolve(trackingsMock);
    });
  });

  it("snapshot renders", () => {

    const component = renderer.create(
      <Provider store={store}>
        <HeatMap />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should exist the vehicle on FormSelect options", async () => {
    const wrapper = mount(
      <Provider store={store}>
        <HeatMap />
      </Provider>
    );

    await flushPromises();
    wrapper.update();

    expect(wrapper.find('option')).toHaveLength(2);
    expect(wrapper.find('option').at(1).text()).toBe("Frod Fiesta - AA111AA");
  });

  it("should get trackings of selected vehicle", async () => {
    const wrapper = mount(
      <Provider store={store}>
        <HeatMap />
      </Provider>
    );

    await flushPromises();
    wrapper.update();

    expect(wrapper.find('CardBody').props().children.props.trackings).toHaveLength(0);

    const mockedEvent = {
      target: {
        value: 1
      },
      preventDefault: () => {},
      persist: () => {}
    }
    wrapper.find('FormSelect').props().onChange(mockedEvent);

    jest.runAllTimers();
    await flushPromises();
    wrapper.update();

    //Map should have 6 marquers
    debugger
    expect(wrapper.find('CardBody').props().children.props.trackings).toHaveLength(trackingsMock.data.length);
  });
});