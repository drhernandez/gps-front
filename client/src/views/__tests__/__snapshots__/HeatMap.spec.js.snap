// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeatMap component snapshot renders 1`] = `
<div
  className="main-content-container px-4 container-fluid"
>
  <div
    className="page-header py-4 no-gutters row"
  >
    <div
      className="mb-3 col-sm-12 text-center text-lg-left mb-sm-0 col-12 col-sm-4"
    >
      <span
        className="text-uppercase page-subtitle"
      >
        Mapa
      </span>
      <h3
        className="page-title"
      >
        Histórico
      </h3>
    </div>
    <div
      className="col"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="mb-4 card card-small"
      >
        <div
          className="card-header"
        >
          <form
            className=""
          >
            <div
              className="row"
            >
              <div
                className="col-12 col-lg-5 mb-1 col"
              >
                <select
                  className="form-control custom-select"
                  defaultValue="default"
                  id="feInputState"
                  onChange={[Function]}
                >
                  <option
                    disabled={true}
                    value="default"
                  >
                    Elija un vehículo...
                  </option>
                </select>
              </div>
              <div
                className="col-12 col-lg-4 d-flex mb-1 col"
                style={
                  Object {
                    "maxHeight": "35px",
                  }
                }
              >
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      className="form-control"
                      disabled={true}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="Desde"
                      readOnly={false}
                      type="text"
                      value="09/07/2020"
                    />
                  </div>
                </div>
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      className="form-control"
                      disabled={true}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="Hasta"
                      readOnly={false}
                      type="text"
                      value="09/08/2020"
                    />
                  </div>
                </div>
                <span
                  className="input-group-text"
                  style={
                    Object {
                      "fontSize": "1rem!important",
                    }
                  }
                >
                  <i
                    className="material-icons"
                  >
                    
                  </i>
                </span>
              </div>
              <div
                className="col-12 col-lg-2 col"
              >
                <button
                  className="update-btn btn btn-accent btn-block"
                  disabled={true}
                  onClick={[Function]}
                  type="button"
                >
                  Actualizar
                </button>
              </div>
            </div>
          </form>
        </div>
        <div
          className="card-body"
        >
          <div
            style={
              Object {
                "height": "100%",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
